pkg/sources/gcs/gcs.go

```diff
@@ -34,27 +57,58 @@ type Source struct {
 	log        logr.Logger
 	chunksCh   chan *sources.Chunk
 
-	mu               sync.Mutex
-	sources.Progress // not safe for concurrent access.
+	processedObjects int32
+	cacheMgr         *cacheManager
+
+	sources.Progress
 }
 
-// Ensure the Source satisfies the interface at compile time.
-var _ sources.Source = (*Source)(nil)
+// cacheManager handles all cache operations with some additional bookkeeping.
+// The threshold value is the percentage of objects that must be processed
+// before the cache is persisted.
+type cacheManager struct {
+	// shouldCache is true if the threshold is > 1.
+	// This ensures that only jobs with a large number of objects will
+	// benefit from caching.
+	shouldCache bool
+	threshold   int
+	cache       cache.Cache
+	progress    *sources.Progress
+}
 
-// Type returns the type of source.
-// It is used for matching source types in configuration and job input.
-func (s *Source) Type() sourcespb.SourceType {
-	return sourcespb.SourceType_SOURCE_TYPE_GCS
+// newCacheManager creates the cache manager with the given threshold.
+func newCacheManager(threshold int, cache cache.Cache, p *sources.Progress) *cacheManager {
+	// If the threshold is <= 1, there is not benefit to caching.
+	shouldCache := threshold > 1
+	return &cacheManager{
+		shouldCache: shouldCache,
+		threshold:   threshold,
+		cache:       cache,
+		progress:    p,
+	}
 }
 
-// SourceID number for GCS Source.
-func (s *Source) SourceID() int64 {
-	return s.sourceId
+func (c *cacheManager) exists(key string) bool {
+	return c.cache.Exists(key)
 }
 
-// JobID number for GCS Source.
-func (s *Source) JobID() int64 {
-	return s.jobId
+func (c *cacheManager) set(key string) {
+	c.cache.Set(key, key)
+}
+
+func (c *cacheManager) shouldPersist() (bool, string) {
+	if !c.shouldCache {
+		return false, ""
+	}
+
+	if c.cache.Count()%c.threshold != 0 {
```

Comment:
I figured it would be an easy way to determine if the number of items in the cache is a multiple of our threshold. If it is we can then persist the data in the cache to the DB within the `Progress`'s `EncodeResumeInfo` field.

I'd love to get your thoughts if you had an alternative approach in mind.